pd/src/state.rs

```diff
@@ -181,4 +184,61 @@ INSERT INTO notes (
         .await?;
         Ok(TransactionDetail { id: id.0 })
     }
+
+    /// Retrieve the [`Asset`] for a given asset ID.
+    pub async fn asset_lookup(&self, asset_id: Vec<u8>) -> Result<Asset> {
+        let mut conn = self.pool.acquire().await?;
+
+        let asset = query_as::<_, (String, Vec<u8>)>(
+            "SELECT denom, asset_id FROM assets WHERE asset_id = $1",
+        )
+        .bind(asset_id)
+        .fetch_one(&mut conn)
+        .await?;
+        Ok(Asset {
+            asset_denom: asset.0,
+            asset_id: asset.1,
+        })
+    }
+
+    /// Retrieves the entire Asset Registry.
+    pub async fn asset_list(&self) -> Result<Vec<Asset>> {
+        let mut conn = self.pool.acquire().await?;
+
+        Ok(
+            query_as::<_, (String, Vec<u8>)>("SELECT denom, asset_id FROM assets")
+                .fetch_all(&mut conn)
+                .await?
+                .into_iter()
+                .map(|(asset_denom, id)| Asset {
+                    asset_denom,
+                    asset_id: id,
+                })
+                .collect(),
+        )
+    }
+
+    /// Given a set of assets, saves them to the Asset Registry database table.
+    pub async fn save_assets_to_registry(
+        &self,
+        assets: &BTreeMap<asset::Id, String>,
+    ) -> Result<()> {
+        let mut dbtx = self.pool.begin().await?;
+
+        for asset in assets {
+            query(
+                r#"
+    INSERT INTO assets (
+        asset_id, denom
+    ) VALUES ($1, $2)
+    "#,
+            )
+            .bind(&asset.0.to_bytes()[..])
+            .bind(asset.1)
+            .execute(&mut dbtx)
+            .await?;
+        }
+
+        dbtx.commit().await.map_err(Into::into)
```

Comment:
Instead of having this logic be a standalone function, I think it would be better if it were part of `commit_block`, which is already going to be processing the `PendingBlock` -- that way, we can ensure that every database write happens as part of one big atomic transaction.

Reply:
Or should it accept a transaction, that way we can re-use it to save assets whenever new ones are found in other blocks?